go/libraries/doltcore/sqle/kvexec/merge_join.go

```diff
@@ -0,0 +1,498 @@
+	rightKey  val.Tuple
+	rightVal  val.Tuple
+
+	lrCmp func(val.Tuple, val.Tuple, val.Tuple, val.Tuple) int
+	llCmp func(val.Tuple, val.Tuple, val.Tuple, val.Tuple) int
+
+	// Non-covering secondary index reads are faster separated
+	// from the initial mapIter, because we avoid unnecessary
+	// primary reads. Note: keyless indexes are less amenable to
+	// this optimization because their cardinality is stored in the
+	// primary index.
+	leftNorm  coveringNormalizer
+	rightNorm coveringNormalizer
+
+	nextRightKey val.Tuple
+	nextRightVal val.Tuple
+
+	lookaheadBuf [][]byte
+	matchPos     int
+
+	// projections
+	joiner *prollyToSqlJoiner
+
+	// todo: we want to build KV-side static expression implementations
+	// so that we can execute filters more efficiently
+	joinFilters []sql.Expression
+	leftFilter  sql.Expression
+	rightFilter sql.Expression
+
+	// LEFT_JOIN impl details
+	excludeNulls bool
+	isLeftJoin   bool
+	matchedLeft  bool
+	exhaustLeft  bool
+}
+
+var _ sql.RowIter = (*mergeJoinKvIter)(nil)
+
+func (l *mergeJoinKvIter) Close(_ *sql.Context) error {
+	return nil
+}
+
+func (l *mergeJoinKvIter) Next(ctx *sql.Context) (sql.Row, error) {
+	var err error
+	if l.leftKey == nil {
+		if err := l.initialize(ctx); err != nil {
+			if errors.Is(err, io.EOF) && l.isLeftJoin {
+				goto exhaustLeft
+			}
+			return nil, err
+		}
+	}
+
+	if l.exhaustLeft {
+		goto exhaustLeft
+	}
+
+	if len(l.lookaheadBuf) > 0 || l.matchPos > 0 {
+		goto match
+	}
+
+incr:
+	// compare the left/right keys.
+	// if equal continue to match buffer stage
+	// if unequal increment one side
+	switch l.lrCmp(l.leftKey, l.leftVal, l.rightKey, l.rightVal) {
+	case -1:
+		// left side has to consider left join non-matches
+		var oldLeftKey, oldLeftVal val.Tuple
+		{
+			// left join state
+			if !l.matchedLeft && l.isLeftJoin {
+				oldLeftKey, oldLeftVal = l.leftKey, l.leftVal
+			}
+			l.matchedLeft = false
+		}
+
+		l.leftKey, l.leftVal, err = l.leftIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, io.EOF) && oldLeftKey != nil {
+				l.exhaustLeft = true
+				candidate, ok, err := l.tryReturn(ctx, oldLeftKey, oldLeftVal, nil, nil)
+				if err != nil {
+					return nil, err
+				}
+				if ok {
+					return candidate, nil
+				}
+			}
+			return nil, err
+		}
+
+		if oldLeftKey != nil {
+			candidate, ok, err := l.tryReturn(ctx, oldLeftKey, oldLeftVal, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				return candidate, nil
+			}
+		}
+	case 0:
+		goto matchBuf
+	case +1:
+		// right side considers lookahead used for match buffer
+		if l.nextRightKey != nil {
+			l.rightKey, l.rightVal = l.nextRightKey, l.nextRightVal
+			l.nextRightKey = nil
+		} else {
+			l.rightKey, l.rightVal, err = l.rightIter.Next(ctx)
+			if err != nil {
+				if errors.Is(err, io.EOF) && l.isLeftJoin {
+					goto exhaustLeft
+				}
+				return nil, err
+			}
+		}
+	}
+	if l.leftKey == nil || l.rightKey == nil {
+		return nil, io.EOF
+	}
+	goto incr
+
+matchBuf:
```

Comment:
Seems like this operation would be more naturally expressed as a loop, even just a naked `for {`